src/Stratis.Bitcoin/Base/AsyncWork/AsyncProvider.cs

```diff
@@ -0,0 +1,212 @@
+                    return;
+                }
+
+                this.logger.LogError(task.Exception.GetBaseException(), "Unhandled exception for async delegate worker {0}", delegateInfo.FriendlyName);
+            }
+        }
+
+        /// <summary>
+        ///  This method is called when delegateTask completed or was canceled.
+        ///  It removes the delegate worker from the dictionary.
+        /// </summary>
+        /// <param name="task">The delegate task.</param>
+        /// <param name="state">The IAsyncDelegate that's run by the delegateTask</param>
+        private void onAsyncDelegateCompleted(Task task, object state)
+        {
+            bool removed;
+            lock (this.lockAsyncDelegates)
+            {
+                removed = this.asyncDelegateWorkers.Remove((IAsyncDelegateDequeuer)state);
+            }
+
+            if (removed)
+            {
+                this.logger.LogTrace("Async delegate worker task removed. Id: {0}", task.Id);
+            }
+            else
+            {
+                // Should never happen.
+                this.logger.LogError("Cannot find the async delegate worker task with Id {0}", task.Id);
+            }
+        }
+
+        /// <inheritdoc />
+        public IAsyncLoop CreateAndRunAsyncLoop<T>(string name, Func<CancellationToken, Task> loop, CancellationToken cancellation, TimeSpan? repeatEvery = null, TimeSpan? startAfter = null)
+        {
+            Guard.NotEmpty(name, nameof(name));
+            Guard.NotNull(loop, nameof(loop));
+            Guard.NotNull(cancellation, nameof(cancellation));
+
+            // instantiate the loop
+            IAsyncLoop loopInstance = new AsyncLoop(name, this.logger, loop);
+
+            Task loopTask;
+            lock (this.lockAsyncLoops)
+            {
+                this.asyncLoops.Add(loopInstance, new AsyncTaskInfo(name));
+            }
+
+            loopTask = loopInstance.Run(cancellation, repeatEvery ?? TimeSpan.FromMilliseconds(DefaultLoopRepeatInterval), startAfter).RunningTask;
+
+            // task will continue with onAsyncDelegateUnhandledException if @delegate had unhandled exceptions
+            loopTask.ContinueWith(this.onAsyncLoopUnhandledException, loopInstance, TaskContinuationOptions.OnlyOnFaulted);
+
+            // task will continue with onAsyncDelegateCompleted if @delegate completed or was canceled
+            loopTask.ContinueWith(this.onAsyncLoopCompleted, loopInstance, TaskContinuationOptions.NotOnFaulted);
+
+            return loopInstance;
+        }
+
+        /// <summary>
+        ///  This method is called when delegateTask had an unhandled exception.
+        /// </summary>
+        /// <param name="task">The loop task.</param>
+        /// <param name="state">The <see cref="IAsyncDelegateDequeuer"/> that's run by the delegateTask</param>
+        private void onAsyncLoopUnhandledException(Task task, object state)
```

Comment:
I wonder can we not reus ethe same continuation as the queue?

Reply:
I thought that but actually I'm storing stuffs in 2 different dictionaries loops and log different message, at the end the duplication is just the population of IAsyncTaskInfoSetter (and I can move that in a private helper)

Reply:
🤔 that's why I proposed using a single dictionary, do we really need to separate the collections? you always have the mandatory friendly name and type to distinguish instances.

Reply:
well they are actually two different objects/interfaces.

The reason of having 2 dictionaries was to print out differently loops and queue infos but I could have a common dictionary, I don't mind

The point is that we may want to handle those "events" differently based on type, I don't think we should handle loops and AsyncDelegate the same way, otherwise why did we need to have 2 different classes?
Until we want to merge AsyncLoop with AsyncDelegate I don't see why we should handle them the same way 

We should talk about this

Reply:
Ok I am convinced, especially that its not much more code to have separate collections.